### PR TITLE
Add check to ensure workload compliance with the next k8s version

### PR DESCRIPTION
**Pull request by tkrishtop**
build-depends: https://github.com/redhatci/ansible-collection-redhatci-ocp/pull/421

Test setup: 

OCP 4.11.18 running as a local CRC-2.12.
On the top, deploying [eventtest-operator](https://github.com/redhatci/eventtest-operator?tab=readme-ov-file#demo-2-on-crc-2120-41118) using API [events.k8s.io/v1beta1](https://github.com/redhatci/eventtest-operator/blob/main/scripts/send_events.py#L39-L40) deprecated in OCP-4.12 / k8s 1.25. This API is used dynamically, out of Python kubernetes package. 

```
$ oc version
Client Version: 4.11.18
Kustomize Version: v4.5.4
Server Version: 4.11.18
Kubernetes Version: v1.24.6+5658434

$ make show
kubectl get crds | grep event
eventtests.eventtest.com                                          2024-07-31T12:10:01Z
kubectl get eventtests -n default
NAME           AGE
my-eventtest   30s
kubectl get pods
NAME                                 READY   STATUS    RESTARTS   AGE
eventtest-operator-978d77ddb-hh5cg   1/1     Running   0          28s

# eventtest-operator is running and sending events
$ oc get events | grep soon
21s         Normal    BreakingNews                                 eventtest/my-eventtest                    Rumors report messages sent to null. More updates expected soon. Stay tuned!
2s          Normal    BreakingNews                                 eventtest/my-eventtest                    Rumors report messages sent to null. More updates expected soon. Stay tuned!
12s         Normal    BreakingNews                                 eventtest/my-eventtest                    Rumors report messages sent to null. More updates expected soon. Stay tuned!
```

CNF config and invocation:

```
$ cat $ cat krishtop_config/tnf_config.yml
---
targetNameSpaces:
  - name: default

podsUnderTestLabels:
  - "name: eventtest-operator"

operatorsUnderTestLabels:
  - "name: eventtest-operator"


$ ./certsuite run --label-filter 'observability' --config-file krishtop_config/tnf_config.yml --kubeconfig= ~/.kube/config --output-dir results

-----------------------------------------------------------
| LOG (observability-compatibility-with-next-ocp-release) |
-----------------------------------------------------------
[INFO] [Sep  5 01:46:06.989] [check.go: 270] [observability-compatibility-with-next-ocp-release] Running check (labels: [common observability-compatibility-with-next-ocp-release observability])
[INFO] [Sep  5 01:46:06.989] [suite.go: 397] [observability-compatibility-with-next-ocp-release] Is OCP: true
[INFO] [Sep  5 01:46:07.888] [suite.go: 414] [observability-compatibility-with-next-ocp-release] 
Detected 4 unique service account names for the workload: 
map[builder:{} default:{} deployer:{} eventtest-operator-service-account:{}]
[INFO] [Sep  5 01:46:07.888] [checksdb.go: 115] [observability-compatibility-with-next-ocp-release] 
Recording result "FAILED", 
claimID: {Id:observability-compatibility-with-next-ocp-release Suite:observability Tags:common}
```

claim.json

```
      "observability-compatibility-with-next-ocp-release": {
        "capturedTestOutput": "[INFO] [Sep  6 06:15:59.009] [check.go: 270] [observability-compatibility-with-next-ocp-release] Running check (labels: [common observability-compatibility-with-next-ocp-release observability])\n[INFO] [Sep  6 06:15:59.009] [suite.go: 398] [observability-compatibility-with-next-ocp-release] Is OCP: true\n[INFO] [Sep  6 06:15:59.658] [suite.go: 415] [observability-compatibility-with-next-ocp-release] Detected 4 unique service account names for the workload: map[builder:{} default:{} deployer:{} eventtest-operator-service-account:{}]\n[INFO] [Sep  6 06:15:59.659] [checksdb.go: 115] [observability-compatibility-with-next-ocp-release] Recording result \"FAILED\", claimID: {Id:observability-compatibility-with-next-ocp-release Suite:observability Tags:common}\n",
        "catalogInfo": {
          "bestPracticeReference": "https://redhat-best-practices-for-k8s.github.io/guide/#redhat-best-practices-for-k8s-to-be-removed-apis",
          "description": "Checks to ensure if the APIs the workload uses are compatible with the next OCP version",
          "exceptionProcess": "No exceptions",
          "remediation": "Ensure the APIs the workload uses are compatible with the next OCP version"
        },
        "categoryClassification": {
          "Extended": "Optional",
          "FarEdge": "Optional",
          "NonTelco": "Optional",
          "Telco": "Optional"
        },
        "checkDetails": "{\"CompliantObjectsOut\":[{\"ObjectType\":\"API\",\"ObjectFieldsKeys\":
[\"Reason For Compliance\",\"ActiveInRelease\",\"APIName\",
\"ServiceAccount\"],
\"ObjectFieldsValues\":[\"API flowschemas.v1beta1.flowcontrol.apiserver.k8s.io used 
by service account default is compliant with Kubernetes version 1.25.0, 
it will be removed in release 1.26\",\"1.25.0\",
\"flowschemas.v1beta1.flowcontrol.apiserver.k8s.io\",\"default\"]},
{\"ObjectType\":\"API\",\"ObjectFieldsKeys\":
[\"Reason For Compliance\",\"ActiveInRelease\",\"APIName\",
\"ServiceAccount\"],\"ObjectFieldsValues\":
[\"API prioritylevelconfigurations.v1beta1.flowcontrol.apiserver.k8s.io used by service account 
default is compliant with Kubernetes version 1.25.0,
 it will be removed in release 1.26\",\"1.25.0\",\"prioritylevelconfigurations.v1beta1.flowcontrol.apiserver.k8s.io\",
\"default\"]}],
\"NonCompliantObjectsOut\":
[{\"ObjectType\":\"API\",\"ObjectFieldsKeys\":
[\"Reason For Non Compliance\",\"RemovedInRelease\",\"APIName\",
\"ServiceAccount\"],\"ObjectFieldsValues\":
[\"API events.v1beta1.events.k8s.io used by service account eventtest-operator-service-account 
is NOT compliant with Kubernetes version 1.25.0, 
it will be removed in release 1.25\",\"1.25\",\"events.v1beta1.events.k8s.io\",
\"eventtest-operator-service-account\"]}]}",
        "duration": 0,
        "endTime": "2024-09-06 06:15:59.659116808 +0200 CEST m=+2.662560049",
        "failureLineContent": "",
        "failureLocation": "",
        "skipReason": "",
        "startTime": "2024-09-06 06:15:59.0095167 +0200 CEST m=+2.012959932",
        "state": "failed",
        "testID": {
          "id": "observability-compatibility-with-next-ocp-release",
          "suite": "observability",
          "tags": "common"
        }
      },
```

### CODE REVIEW

**Comment by sebrandon1** [comment]
The test setup you listed could eventually be added to our QE repo as well to test positive and negative cases:
https://github.com/test-network-function/cnfcert-tests-verification

Our PR testers here in this repo only run `observability` against k8s at the moment but could be adjust in the future to run against OCP as well.

**Comment by tkrishtop** [comment]
@sebrandon1 could you please allow the tests to run or grant me the permissions to do so? Thank you!

**Comment by sebrandon1** [comment]
Approved!

**Comment by tkrishtop** [comment]
Hi @sebrandon1 could you please approve the workloads once again to see if my fixes have worked?  
I've also set the PR ready for the review, would appreciate your feedback.

**Comment by tkrishtop** [comment]
> Hi @sebrandon1 could you please approve the workloads once again to see if my fixes have worked? I've also set the PR ready for the review, would appreciate your feedback.

@sebrandon1 sorry, forgot to rename to best_practices in one file. Could you please re-approve the workloads?

**Comment by tkrishtop** [comment]
Linter and unittests now work on my laptop:
- Linter

```
$ podman run --rm -v $(pwd):/app:z -w /app --user root golangci/golangci-lint:v1.59 golangci-lint run --timeout 10m0s --config /app/.golangci.yml -v
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
level=info msg="golangci-lint has version 1.59.1 built with go1.22.3 from 1a55854a on 2024-06-09T18:08:33Z"
level=info msg="[config_reader] Used config file .golangci.yml"
level=info msg="[lintersdb] Active 27 linters: [bodyclose dogsled errcheck exhaustive exportloopref funlen goconst gocritic gocyclo gofmt goimports goprintffuncname gosec gosimple govet ineffassign lll misspell mnd nolintlint rowserrcheck staticcheck stylecheck unconvert unparam unused whitespace]"
level=info msg="[loader] Go packages loading at mode 575 (name|deps|exports_file|files|imports|types_sizes|compiled_files) took 50.840948537s"
level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 17.146554ms"
level=info msg="[linters_context/goanalysis] analyzers took 10m58.386230029s with top 10 stages: buildir: 5m7.266472s, gocritic: 4m14.274573648s, nilness: 11.70157287s, exhaustive: 8.650687668s, unparam: 7.314610456s, buildssa: 6.76775657s, inspect: 6.059644581s, fact_purity: 5.771360406s, printf: 4.922116336s, unconvert: 4.877593235s"
level=info msg="[runner] Issues before processing: 491, after processing: 0"
level=info msg="[runner] Processors filtering stat (out/in): exclude-rules: 211/489, exclude: 489/489, invalid_issue: 491/491, autogenerated_exclude: 489/491, nolint: 0/211, filename_unadjuster: 491/491, skip_dirs: 491/491, skip_files: 491/491, path_prettifier: 491/491, identifier_marker: 489/489, cgo: 491/491"
level=info msg="[runner] processing took 74.952742ms with stages: nolint: 41.515541ms, identifier_marker: 13.958188ms, exclude-rules: 13.800695ms, autogenerated_exclude: 2.929506ms, path_prettifier: 1.788674ms, skip_dirs: 735.803µs, cgo: 122.939µs, invalid_issue: 58.242µs, filename_unadjuster: 39.01µs, max_same_issues: 996ns, exclude: 491ns, uniq_by_line: 418ns, fixer: 383ns, skip_files: 376ns, max_from_linter: 324ns, sort_results: 258ns, source_code: 238ns, diff: 207ns, max_per_file_from_linter: 124ns, path_shortener: 119ns, path_prefixer: 108ns, severity-rules: 102ns"
level=info msg="[runner] linters took 57.144779393s with stages: goanalysis_metalinter: 57.069704462s"
level=info msg="File cache stats: 277 entries of total size 1.5MiB"
level=info msg="Memory: 967 samples, avg is 1597.2MB, max is 6081.3MB"
level=info msg="Execution took 1m48.010969915s"
```

@sebrandon1 could you please approve the workloads once again?

**Review by sebrandon1 — APPROVED** [review]
This looks good.  We'll have to get some QE tests setup to do some positive/negative tests on OCP.

**Comment by tkrishtop** [comment]
> This looks good. We'll have to get some QE tests setup to do some positive/negative tests on OCP.

Hi @sebrandon1 thank you for the review! If that's fine, I'll schedule a 15-minutes sync with you next week to discuss on how to proceed with test scenarios to add them into your [QE repo](https://github.com/redhat-best-practices-for-k8s/certsuite-qe).

**Review by greyerof — COMMENTED** [review]
Thanks for your contribution @tkrishtop !

I left some comments, but I also have a couple of concerns:
- The "[byUser](https://github.com/redhat-best-practices-for-k8s/certsuite/pull/2303/files#diff-acdf00ff10eb8af031e9830e54b7748334ec7e5c0931f36c8b07a663f7dd7ef1R301)" iteration method. As per my investigation, the reports of APIRequestCount is limited to 10 users, and they seem ordered by number of requests, so we might not get requests for deprecated APIs from a workload if it doesn't reach the minimum number of requests as to appear in that list. This users number can be updated/increased (patching the APIRequestCount object), but I'm not sure if that's a good way to go as we would be modifying the cluster.
- The workload under test might not be running or it might not have requested the deprecated API yet, thus we won't catch it.

Unless I'm wrong or I'm missing something, this means that a "pass" in this TC might be misleading, or simply wrong, so I'm not sure if this is a good test case to add :(

**Comment by tkrishtop** [comment]
> * As per my investigation, the reports of APIRequestCount is limited to 10 users

Hi @greyerof, thank you for the extensive review — I really appreciate it. Could you please provide more information about the 10-user limit you mentioned? Are you referring to the number of service accounts per API call when checking the currentHour structure? Please note that here we are dealing with last24h, which generally contains more information.

**Comment by greyerof** [comment]
@tkrishtop As per our official [docs](https://docs.openshift.com/container-platform/4.14/rest_api/metadata_apis/apirequestcount-apiserver-openshift-io-v1.html), it's defaulted to 10.

image

This seems to be consistent with what I can get from a 4.15.3 SNO cluster:
```
$  oc get apirequestcounts.apiserver.openshift.io -o json | jq '[ .items[] | { "api-name": .metadata.name, "numberOfUsersToReport": .spec.numberOfUsersToReport } ]'
[
  {
    "api-name": "adminpolicybasedexternalroutes.v1.k8s.ovn.org",
    "numberOfUsersToReport": 10
  },
  {
    "api-name": "alertingrules.v1.monitoring.openshift.io",
    "numberOfUsersToReport": 10
  },
  {
    "api-name": "alertmanagerconfigs.v1alpha1.monitoring.coreos.com",
    "numberOfUsersToReport": 10
  },
  {
    "api-name": "alertmanagerconfigs.v1beta1.monitoring.coreos.com",
    "numberOfUsersToReport": 10
  },
  ...
```
Regarding the service account: it seems it's reporting accesses by service account, so it seems possible to filter the api requests to check only the ones that our workload-under-test is actually using.

image

**Comment by greyerof** [comment]
@tkrishtop The field `last24h` shows only the users with the (descending order) greatest number of requets. E.g.:

For "pods.v1" in the first hour for the first node:
```
$ oc get apirequestcounts.apiserver.openshift.io pods.v1 -o json | jq '.status.last24h[0] | .byNode[0] | .byUser | length' 
10
```

Requests by user on the first hour in the first node:
```
$ oc get apirequestcounts.apiserver.openshift.io pods.v1 -o json | jq '[ .status.last24h[0] | .byNode[0] | .byUser[] | { "user" : .userAgent, "requestCount": .requestCount } ]' 
[
  {
    "user": "kubelet/v1.28.7+6e2789b",
    "requestCount": 513
  },
  {
    "user": "Prometheus/2.48.0",
    "requestCount": 357
  },
  {
    "user": "cluster-kube-apiserver-operator/v0.0.0",
    "requestCount": 200
  },
  {
    "user": "cluster-kube-scheduler-operator/v0.0.0",
    "requestCount": 177
  },
  {
    "user": "cluster-kube-controller-manager-operator/v0.0.0",
    "requestCount": 177
  },
  {
    "user": "cluster-etcd-operator/v0.0.0",
    "requestCount": 175
  },
  {
    "user": "multus-daemon/v0.0.0",
    "requestCount": 141
  },
  {
    "user": "catalog/v0.0.0",
    "requestCount": 83
  },
  {
    "user": "authentication-operator/v0.0.0",
    "requestCount": 66
  },
  {
    "user": "ingress-operator/v0.0.0",
    "requestCount": 46
  }
]
```

**Comment by tkrishtop** [comment]
> the reports of APIRequestCount is limited to 10 users. -- snip -- This users number can be updated/increased

> The workload under test might not be running or it might not have requested the deprecated API yet, thus we won't catch it

> Unless I'm wrong or I'm missing something, this means that a "pass" in this TC might be misleading, or simply wrong, so I'm not sure if this is a good test case to add :(

Hi @greyerof, 

Regarding `numberOfUsersToReport`, it is not a blocker and can be adjusted if needed. However, I don’t believe there will be a real need to adjust it in practice, as it’s uncommon to find beta or alpha APIs being used by more than 10 service accounts simultaneously. The example you provided for `pods.v1` is out of scope; `pods.v1` is GA and heavily used by internal Kubernetes services, which is not the type of to-be-deprecated API we’re focusing on here. In my experience working with Telco partners, I’ve never encountered such a situation with non-GA alpha or beta APIs. But even if it does exist, there’s a way to address it.

So, the key question here is: Is it worthwhile to add this check, knowing that the usage of `APIRequestCount` requires a workload to be running/passing integration tests first? I believe it is. We’ve been running this check out of DCI for about a year and have several partners using it heavily. It’s currently the only approach on the market that detects dynamically invoked APIs, such as those called from scripts or via the Go/Python Kubernetes packages. Other solutions only check static invocations by parsing manifests.

With one of our partners, Keysight, we presented this solution at [DevConf2024](https://pretalx.com/devconf-cz-2024/talk/review/UAHGESHZNM7BPSZ9UUQUYABDZWRNRLJF) and received extremely positive feedback from the MongoDB team—they face the same problem with every Kubernetes upgrade. That’s why we proposed this solution as part of the Red Hat Best Practices Certification for Kubernetes.

That said, I’m not a maintainer, so the strategy moving forward is up to you and @sebrandon1. Please let me know if you’d prefer to have a meeting to discuss this further. Thanks!

File: tests/observability/suite.go
Comment (edcdavid) [reply tests/observability/suite.go:302]: yes I agree, would be better to track requests from the workload under test only. I wonder if we could miss some api calls that way though. Probably will require some more testing. If this is informational/optional, I am not opposed to have a more accurate result in a future PR.

Comment (edcdavid) [reply tests/observability/suite.go:302]: Looks like the original ansible project from the video also scope the results via namespace for similar reasons. @tkrishtop

**Comment by greyerof** [comment]
Hi, @tkrishtop . Thanks for your clarifications, now I understand it. 

It's ok to include this test-case/check as part of the certsuite, but we believe it should only fail if any of the "to be removed" api requests were actually requested by the workload-under-test. Anyway, this filter can be done in a futher PR.

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go]: thank you, good idea, done

File: pkg/provider/provider.go
Comment (tkrishtop) [reply pkg/provider/provider.go]: thanks, done

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go]: done

File: pkg/provider/provider.go
Comment (tkrishtop) [reply pkg/provider/provider.go]: thank you for the hint, switched to the way you have proposed

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go]: done, thank you for the example

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go]: good point, thank you, updated the test results to look like this
```
"checkDetails": "{\"CompliantObjectsOut\":null,
\"NonCompliantObjectsOut\":
[{\"ObjectType\":\"API\",\"ObjectFieldsKeys\":
[\"Reason For Non Compliance\",
\"RemovedInRelease\",\"APIName\",\"UserInfo\",
\"UserInfo\"],\"ObjectFieldsValues\":
[\"API is not compliant with Kubernetes version 1.25.0 
because it will be removed in release 1.25\",\"1.25\",
\"events.v1beta1.events.k8s.io\",
\"UserName: system:serviceaccount:default:eventtest-operator-service-account, 
UserAgent: OpenAPI-Generator/24.2.0/python\",
\"UserName: system:serviceaccount:default:eventtest-operator-service-account, 
UserAgent: ansible-operator/v0.0.0\"]}]}",
```

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go:334]: @greyerof I see your point, but I’m concerned that for real workloads, especially in Telcos, we might end up with far too many APIs to display. The standard size of an APIRequestCount object per namespace for Telco workloads is about 100MB, and for one partner, we’ve observed 22 workspaces producing 22GB of APIRequestCount data. That’s why I’ve started filtering out everything:

- Keeping only APIs with a non-zero removedInRelease field
- Removing service calls with service accounts starting with openshift* or kube*
- Checking APIs per namespace to reduce the overall size of the processed data

In the comment below, I see that in CNF you have the option to consider only workload-related APIs by checking the workload's ServiceAccounts through env.testPods. I’ll explore this option.

**Comment by tkrishtop** [comment]
Hi @greyerof,

I applied all the recommendations you suggested, and thank you once again for taking the time to provide such a detailed review. I have two points to highlight:

1. I used `env.ServiceAccounts`, which is built from `json:"testServiceAccounts"`, to filter and keep only workload-related APIs.
2. If the workload is not using any to-be-deprecated APIs and is only using GA APIs, I simply add a stub object to make the test pass.

Could you please take another look?

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go]: I used env.ServiceAccounts, which is built from json:"testServiceAccounts", to filter and keep only workload-related APIs. @greyerof @edcdavid

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go:339]: If the workload is not using any to-be-deprecated APIs and is only using GA APIs, I added a stub object to make the test pass.

**Comment by tkrishtop** [comment]
retest

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go]: Hi @greyerof I like your idea, just added the release version in which it will be removed as the key
```
// Define a map where the key is the service account name and the value is another map
// The inner map key is the API name and the value is the release version in which it will be removed
serviceAccountToDeprecatedAPIs := make(map[string]map[string]string)
```

File: tests/observability/suite.go
Comment (tkrishtop) [reply tests/observability/suite.go]: yep, makes sense, added this idea into code as well. Thanks

**Comment by tkrishtop** [comment]
Hi @greyerof I actually liked both ideas; they both make sense, so I integrated them. Thank you.

**Review by greyerof — APPROVED** [review]

**Review by sebrandon1 — APPROVED** [review]

**Pull request merged**
